File: CatchOPP/CommunityMicroService/src/main/java/tn/esprit/communitymicroservice/services/JoinRequestService.java
package tn.esprit.communitymicroservice.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.esprit.communitymicroservice.entities.*;
import tn.esprit.communitymicroservice.repositories.GroupMemberRepository;
import tn.esprit.communitymicroservice.repositories.GroupRepository;
import tn.esprit.communitymicroservice.repositories.JoinRequestRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@Service
public class JoinRequestService {

    @Autowired
    private JoinRequestRepository joinRequestRepository;

    @Autowired
    private GroupRepository groupRepository;

    @Autowired
    private GroupMemberRepository groupMemberRepository;

    // ── Submit a join request ─────────────────────────────────────────────
    public JoinRequest requestJoin(Long groupId, Long userId) {
        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found: " + groupId));

        // Only INVITE_ONLY groups require a request
        if (group.getType() != GroupType.INVITE_ONLY) {
            throw new RuntimeException("Group is not invitation-only. Join directly.");
        }

        // Reject duplicate pending request
        if (joinRequestRepository.existsByGroupIdAndUserIdAndStatus(groupId, userId, JoinRequestStatus.PENDING)) {
            throw new RuntimeException("A pending request already exists for this user and group.");
        }

        JoinRequest request = new JoinRequest();
        request.setGroup(group);
        request.setUserId(userId);
        request.setStatus(JoinRequestStatus.PENDING);
        return joinRequestRepository.save(request);
    }

    // ── Get all requests for a group ──────────────────────────────────────
    public List<JoinRequest> getAllByGroup(Long groupId) {
        return joinRequestRepository.findByGroupId(groupId);
    }

    // ── Get pending requests for a group ──────────────────────────────────
    public List<JoinRequest> getPendingByGroup(Long groupId) {
        return joinRequestRepository.findByGroupIdAndStatus(groupId, JoinRequestStatus.PENDING);
    }

    // ── Get all requests from a user ──────────────────────────────────────
    public List<JoinRequest> getAllByUser(Long userId) {
        return joinRequestRepository.findByUserId(userId);
    }

    // ── Accept a request → add user as member ────────────────────────────
    public JoinRequest acceptRequest(Long requestId) {
        JoinRequest request = joinRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found: " + requestId));

        if (request.getStatus() != JoinRequestStatus.PENDING) {
            throw new RuntimeException("Request is already processed.");
        }

        request.setStatus(JoinRequestStatus.ACCEPTED);
        request.setProcessedAt(LocalDateTime.now());
        joinRequestRepository.save(request);

        // Auto-add as group member
        GroupMember member = new GroupMember();
        member.setGroup(request.getGroup());
        member.setUserId(request.getUserId());
        member.setRole(Role.MEMBER);
        member.setJoinedAt(LocalDateTime.now());
        groupMemberRepository.save(member);

        return request;
    }

    // ── Reject a request ──────────────────────────────────────────────────
    public JoinRequest rejectRequest(Long requestId) {
        JoinRequest request = joinRequestRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found: " + requestId));

        if (request.getStatus() != JoinRequestStatus.PENDING) {
            throw new RuntimeException("Request is already processed.");
        }

        request.setStatus(JoinRequestStatus.REJECTED);
        request.setProcessedAt(LocalDateTime.now());
        return joinRequestRepository.save(request);
    }

    // ── Check if a pending request exists ────────────────────────────────
    public boolean hasPendingRequest(Long groupId, Long userId) {
        return joinRequestRepository.existsByGroupIdAndUserIdAndStatus(
                groupId, userId, JoinRequestStatus.PENDING);
    }

    // ── Get request by ID ─────────────────────────────────────────────────
    public JoinRequest getById(Long id) {
        return joinRequestRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Request not found: " + id));
    }
}

File: FrontFreelanceApp/src/app/interfaces_events/join-request.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_CONFIG } from './api.config';

export interface JoinRequest {
  id: number;
  group: { id: number; name: string };
  userId: number;
  status: 'PENDING' | 'ACCEPTED' | 'REJECTED';
  requestedAt: string;
  processedAt?: string;
}

@Injectable({ providedIn: 'root' })
export class JoinRequestService {
  private apiUrl = `${API_CONFIG.BASE_URL}/api/join-requests`;

  constructor(private http: HttpClient) {}

  requestJoin(groupId: number, userId: number): Observable<JoinRequest> {
    return this.http.post<JoinRequest>(this.apiUrl, { groupId, userId });
  }

  getPendingRequests(groupId: number): Observable<JoinRequest[]> {
    return this.http.get<JoinRequest[]>(`${this.apiUrl}/group/${groupId}/pending`);
  }

  getAllRequests(groupId: number): Observable<JoinRequest[]> {
    return this.http.get<JoinRequest[]>(`${this.apiUrl}/group/${groupId}`);
  }

  acceptRequest(requestId: number): Observable<any> {
    return this.http.put(`${this.apiUrl}/${requestId}/accept`, {});
  }

  rejectRequest(requestId: number): Observable<JoinRequest> {
    return this.http.put<JoinRequest>(`${this.apiUrl}/${requestId}/reject`, {});
  }

  checkPending(groupId: number, userId: number): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/check?groupId=${groupId}&userId=${userId}`);
  }
}
